crates/bevy_scene/src/lib.rs

```diff
@@ -59,6 +59,30 @@ impl Plugin for ScenePlugin {
             .add_event::<SceneInstanceReady>()
             .init_resource::<SceneSpawner>()
             .add_systems(SpawnScene, (scene_spawner, scene_spawner_system).chain());
+
+        // Register component hooks for DynamicScene
+        app.world
+            .register_component_hooks::<Handle<DynamicScene>>()
+            .on_remove(|mut world, entity, _| {
+                let id = world.get::<Handle<DynamicScene>>(entity).unwrap().id();
+                if let Some(&SceneInstance(scene_instance)) = world.get::<SceneInstance>(entity) {
+                    let mut scene_spawner = world.resource_mut::<SceneSpawner>();
+                    if let Some(instance_ids) = scene_spawner.spawned_dynamic_scenes.get_mut(&id) {
+                        instance_ids.remove(&scene_instance);
+                    }
+                    scene_spawner.despawn_instance(scene_instance);
+                }
+            });
+
+        // Register component hooks for Scene
+        app.world
+            .register_component_hooks::<Handle<Scene>>()
+            .on_remove(|mut world, entity, _| {
+                if let Some(&SceneInstance(scene_instance)) = world.get::<SceneInstance>(entity) {
+                    let mut scene_spawner = world.resource_mut::<SceneSpawner>();
```

Comment:
Let's not panic if the resource is not found for whatever reason. If it's missing, there's no reason to be crashing the entire app. Use `world.get_resource_mut` and ignore it if it's `None` for whatever reason.